File: ts/src/ordering.ts
import { OrderingError } from "./errors.js";
import type { SourceBook } from "./models.js";

const SPECIAL_MARKERS = ["bd", "blu-ray", "bluray", "特典", "特装", "限定", "another"];
const CHINESE_DIGITS = new Map<string, number>([
  ["零", 0], ["〇", 0], ["一", 1], ["二", 2], ["三", 3], ["四", 4],
  ["五", 5], ["六", 6], ["七", 7], ["八", 8], ["九", 9]
]);

export function sortSources(sources: SourceBook[]): SourceBook[] {
  const keyed = sources.map((source) => ({ key: orderKey(source), source }));
  const seen = new Map<string, string>();
  for (const item of keyed) {
    const key = item.key.join(".");
    const other = seen.get(key);
    if (other) {
      throw new OrderingError(`Ambiguous input order: '${item.source.basename}' and '${other}' share order key (${item.key.join(",")})`);
    }
    seen.set(key, item.source.basename);
  }
  return keyed.sort((a, b) => compareKey(a.key, b.key)).map((item) => item.source);
}

export function orderKey(source: SourceBook): number[] {
  const filename = source.basename.replace(/\.epub$/i, "");
  const decimal = decimalKey(filename);
  if (decimal) return decimal;
  const special = specialKey(filename);
  if (special) return special;
  for (const text of [filename, source.title, source.toc[0]?.title ?? ""]) {
    const numbers = numberList(text);
    if (numbers.length) return numbers;
  }
  throw new OrderingError(`Could not determine input order for '${source.basename}'`);
}

function compareKey(a: number[], b: number[]): number {
  const length = Math.max(a.length, b.length);
  for (let i = 0; i < length; i++) {
    const av = a[i] ?? -1;
    const bv = b[i] ?? -1;
    if (av !== bv) return av - bv;
  }
  return 0;
}

function decimalKey(text: string): number[] | undefined {
  const match = text.match(/(\d+)\.(\d+)/);
  if (!match) return undefined;
  return [Number(match[1]), Number(match[2].padEnd(6, "0").slice(0, 6))];
}

function specialKey(text: string): number[] | undefined {
  const lowered = text.toLowerCase();
  if (!SPECIAL_MARKERS.some((marker) => lowered.includes(marker))) return undefined;
  const values = [...numberList(text), ...chineseOrdinals(text)].sort((a, b) => a - b);
  if (!values.length) return undefined;
  return [10_000, ...values];
}

function numberList(text: string): number[] {
  return [...text.matchAll(/\d+/g)].map((match) => Number(match[0]));
}

function chineseOrdinals(text: string): number[] {
  return [...text.matchAll(/第([一二三四五六七八九十〇零]+)/g)]
    .map((match) => parseChineseNumber(match[1]))
    .filter((value): value is number => value !== undefined);
}

function parseChineseNumber(text: string): number | undefined {
  if (text === "十") return 10;
  if (text.includes("十")) {
    const [left, right = ""] = text.split("十");
    const tens = left ? CHINESE_DIGITS.get(left) : 1;
    const ones = right ? CHINESE_DIGITS.get(right) : 0;
    if (tens === undefined || ones === undefined) return undefined;
    return tens * 10 + ones;
  }
  let total = 0;
  for (const char of text) {
    const digit = CHINESE_DIGITS.get(char);
    if (digit === undefined) return undefined;
    total = total * 10 + digit;
  }
  return total;
}

File: src/epub_merge_tool/ordering.py
from __future__ import annotations

import re
from pathlib import Path

from .errors import OrderingError
from .models import SourceBook


SPECIAL_MARKERS = ("bd", "blu-ray", "bluray", "特典", "特装", "限定", "another")
CHINESE_DIGITS = {
    "零": 0,
    "〇": 0,
    "一": 1,
    "二": 2,
    "三": 3,
    "四": 4,
    "五": 5,
    "六": 6,
    "七": 7,
    "八": 8,
    "九": 9,
}


def sort_sources(sources: list[SourceBook]) -> list[SourceBook]:
    keyed: list[tuple[tuple[int, ...], SourceBook]] = []
    for source in sources:
        key = order_key(source)
        keyed.append((key, source))
    seen: dict[tuple[int, ...], str] = {}
    for key, source in keyed:
        if key in seen:
            raise OrderingError(
                f"Ambiguous input order: {source.basename!r} and {seen[key]!r} share order key {key}"
            )
        seen[key] = source.basename
    return [source for _, source in sorted(keyed, key=lambda item: item[0])]


def order_key(source: SourceBook) -> tuple[int, ...]:
    filename = Path(source.basename).stem
    decimal = _decimal_key(filename)
    if decimal:
        return decimal
    special = _special_key(filename)
    if special:
        return special
    for text in (filename, source.title, source.toc[0].title if source.toc else ""):
        numbers = _numbers(text)
        if numbers:
            return numbers
    raise OrderingError(f"Could not determine input order for {source.basename!r}")


def _numbers(text: str) -> tuple[int, ...]:
    return tuple(int(match) for match in re.findall(r"\d+", text))


def _decimal_key(text: str) -> tuple[int, ...] | None:
    match = re.search(r"(\d+)\.(\d+)", text)
    if not match:
        return None
    fractional = match.group(2)
    scaled = int(fractional.ljust(6, "0")[:6])
    return int(match.group(1)), scaled


def _special_key(text: str) -> tuple[int, ...] | None:
    lowered = text.lower()
    if not any(marker in lowered for marker in SPECIAL_MARKERS):
        return None
    values = list(_numbers(text))
    values.extend(_chinese_ordinals(text))
    if not values:
        return None
    return (10_000, *sorted(values))


def _chinese_ordinals(text: str) -> tuple[int, ...]:
    values: list[int] = []
    for match in re.findall(r"第([一二三四五六七八九十〇零]+)", text):
        value = _parse_chinese_number(match)
        if value is not None:
            values.append(value)
    return tuple(values)


def _parse_chinese_number(text: str) -> int | None:
    if text == "十":
        return 10
    if "十" in text:
        left, _, right = text.partition("十")
        tens = CHINESE_DIGITS.get(left, 1) if left else 1
        ones = CHINESE_DIGITS.get(right, 0) if right else 0
        return tens * 10 + ones
    total = 0
    for char in text:
        digit = CHINESE_DIGITS.get(char)
        if digit is None:
            return None
        total = total * 10 + digit
    return total
